03-transactions.md

```diff
@@ -1423,6 +1431,26 @@ And, here are the test vectors themselves:
     output commit_tx: 02000000000101bef67e4e2fb9ddeeb3461973cd4c62abb35050b1add772995b820b584a488489000000000038b02b8001c0c62d0000000000160014ccf1af2f2aabee14bb40fa3851ab2301de8431100400473044022031a82b51bd014915fe68928d1abf4b9885353fb896cac10c3fdd88d7f9c7f2e00220716bda819641d2c63e65d3549b6120112e1aeaf1742eed94a471488e79e206b101473044022064901950be922e62cbe3f2ab93de2b99f37cff9fc473e73e394b27f88ef0731d02206d1dfa227527b4df44a07599289e207d6fd9cca60c0365682dcd3deaf739567e01475221023da092f6980e58d2c037173180e9a465476026ee50f96695963e8efe436f54eb21030e9f7b623d2ccc7c9bd44d66d5ce21ce504c0acf6385a132cec6d3c39fa711c152ae3e195220
     num_htlcs: 0
 
+    name: commitment tx with 3 htlc outputs, 2 offered having the same amount and preimage
+    to_local_msat: 6988000000
+    to_remote_msat: 3000000000
+    local_feerate_per_kw: 253
+    # HTLC 0 received amount 1000000 wscript 76a91414011f7254d96b819c76986c277d115efce6f7b58763ac67210394854aa6eab5b2a8122cc726e9dded053a2184d88256816826d6231c068d4a5b7c8201208763a914b8bcb07f6344b42ab04250c86a6e8b75d3fdbbc688527c21030d417a46946384f88d5f3337267c5e579765875dc4daca813e21734b140639e752ae677502f401b175ac6868
+    # HTLC 5 offered amount 5000000 wscript 76a91414011f7254d96b819c76986c277d115efce6f7b58763ac67210394854aa6eab5b2a8122cc726e9dded053a2184d88256816826d6231c068d4a5b7c820120876475527c21030d417a46946384f88d5f3337267c5e579765875dc4daca813e21734b140639e752ae67a9142002cc93ebefbb1b73f0af055dcc27a0b504ad7688ac6868
+    # HTLC 6 offered amount 5000000 wscript 76a91414011f7254d96b819c76986c277d115efce6f7b58763ac67210394854aa6eab5b2a8122cc726e9dded053a2184d88256816826d6231c068d4a5b7c820120876475527c21030d417a46946384f88d5f3337267c5e579765875dc4daca813e21734b140639e752ae67a9142002cc93ebefbb1b73f0af055dcc27a0b504ad7688ac6868
+    # HTLC 5 and 6 have CLTV 505 and 506, respectively, and preimage 0505050505050505050505050505050505050505050505050505050505050505
+    output commit_tx: 02000000000101bef67e4e2fb9ddeeb3461973cd4c62abb35050b1add772995b820b584a488489000000000038b02b8005e80300000000000022002052bfef0479d7b293c27e0f1eb294bea154c63a3294ef092c19af51409bce0e2a8813000000000000220020305c12e1a0bc21e283c131cea1c66d68857d28b7b2fce0a6fbc40c164852121b8813000000000000220020305c12e1a0bc21e283c131cea1c66d68857d28b7b2fce0a6fbc40c164852121bc0c62d0000000000160014ccf1af2f2aabee14bb40fa3851ab2301de843110a79f6a00000000002200204adb4e2f00643db396dd120d4e7dc17625f5f2c11a40d857accc862d6b7dd80e04004730440220586f17b3f3f6eec96a0dc7040f1a33174c707e032010421fa1965497cd1c81e602204476e6a4cc17d433e8ad82edff2ffa47c954ecb52aad9cc34382befb66f6e00001473044022069797b84fad1e0e7b02e1edeec80cb6e0b6f01170700bca273788715f55e0c560220260833550768b1db3ed88f8c005cf2560821b1879ec69834cd9765de25ec350601475221023da092f6980e58d2c037173180e9a465476026ee50f96695963e8efe436f54eb21030e9f7b623d2ccc7c9bd44d66d5ce21ce504c0acf6385a132cec6d3c39fa711c152ae3e195220
```

Comment:
`a79f6a0000000000` is the `to_local` output value, I believe, which is little-endian for 6_987_687. However, in RL I'm getting a value of 6_988_687. I'm seeing `c0c62d0000000000` (3_000_000) for the `to_remote` output value, as expected.

I think this means either our fee calculation is wrong or the test vector is wrong. Could someone verify this or correct my understanding?

Reply:
Interesting, and you diverge only on this test vector?
Doing the calculation by-hand, I find that the commit tx fee should be:

```
commit tx weight = 724
htlc output weight = 172
feerate = 253
fee = (724 + 3 * 172) * 253 / 1000 = 313 sat
```

So `to_local = 6_988_000 - 313 = 6_987_687` which seems to confirm the test vector.
You have exactly 1_000 sat more than that, which is the amount of the incoming HTLC, could you be somehow counting that in your `to_local`?

Reply:
> Interesting, and you diverge only on this test vector?

Yeah, just this one.

> Doing the calculation by-hand, I find that the commit tx fee should be:
> 
> ```
> commit tx weight = 724
> htlc output weight = 172
> feerate = 253
> fee = (724 + 3 * 172) * 253 / 1000 = 313 sat
> ```
> 
> So `to_local = 6_988_000 - 313 = 6_987_687` which seems to confirm the test vector.
> You have exactly 1_000 sat more than that, which is the amount of the incoming HTLC, could you be somehow counting that in your `to_local`?

Ah, I think therein lies the problem! My subconscious churned away at this overnight:

Other than the first test which uses no HTLCs, the other tests use HTLCs 0-4 (according to lines 1013-1014 prior to this PR) which total 12_000_000 msats. I think that is where `to_local_msat: 6988000000` comes from. However, the new test uses HTLCs 0, 5, and 6. These total 11_000_000 msats, which should instead give `to_local_msat: 6989000000`.

Given the PR adds new HTLCs, I think the description in lines 1013-1014 needs to be updated. The language isn't very clear and is no longer accurate with the additional HTLCs.

Reply:
I see what you mean, but I'm not sure it makes sense to interpret the test case this way.

If we consider that we start with `to_local_msat = 7000000000` and `to_remote_msat = 3000000000` and then apply HTLCs 0-4, we should subtract `5000000 msat` from `to_local` and subtract `7000000 msat` from `to_remote`, resulting in `to_local_msat = 6995000000` and `to_remote_msat = 2993000000`.

Another way to read this would be to use the amounts if all HTLCs are fulfilled, but even in that case it doesn't work: we have `7000000 msat` incoming and `5000000 msat` outgoing, so the result would be `to_local_msat = 7002000000` (and `to_remote_msat = 2998000000`).

I can't speak for what the reason was when these test vectors were initially created, but I interpret the `to_local_msat` as simply saying "your main output has this value". I don't think the values of pending HTLCs should play any role here.

> Given the PR adds new HTLCs, I think the description in lines 1013-1014 needs to be updated.

That's right, the two new HTLCs should be described here, it would slightly help read these already hard-to-figure-out tests!

Reply:
> I see what you mean, but I'm not sure it makes sense to interpret the test case this way.
> 
> If we consider that we start with `to_local_msat = 7000000000` and `to_remote_msat = 3000000000` and then apply HTLCs 0-4, we should subtract `5000000 msat` from `to_local` and subtract `7000000 msat` from `to_remote`, resulting in `to_local_msat = 6995000000` and `to_remote_msat = 2993000000`.
> 
> Another way to read this would be to use the amounts if all HTLCs are fulfilled, but even in that case it doesn't work: we have `7000000 msat` incoming and `5000000 msat` outgoing, so the result would be `to_local_msat = 7002000000` (and `to_remote_msat = 2998000000`).
> 
> I can't speak for what the reason was when these test vectors were initially created, but I interpret the `to_local_msat` as simply saying "your main output has this value". I don't think the values of pending HTLCs should play any role here.

While I don't profess to fully understand this myself, the other tests behave in this manner. Namely, for all test vectors, the "output commitment tx" has a `to_remote` output of 3_000_000 (`c0c62d0000000000` little-endian) regardless of which HTLCs are included.

Reply:
Indeed, I view this as an additional reason to decouple the `to_local_msat` value from the HTLCs used (which means the current test vector would be ok).

Can you explain why it's hard to replicate in rust-lightning? Aren't you simply initializing the commitment tx main outputs with the `to_local_msat` and `to_remote_msat` indicated in the test vector?

Reply:
The way the tests (currently) work is we build the full channel state machine and then get a commitment tx from it, not building a commitment tx from raw values. Thus we have to convert the numbers into state machine values (eg there is no amount-to-remote field, to-local tracking doesn't include HTLC values, etc).

Reply:
Got it, but then the HTLC amounts should not impact you, right? If you open a new channel, with the right initial amount (using `push_msat` I guess to have funds on the remote side) and set the `channel_reserve` accordingly, you should be able to have `to_local` and `to_remote` that match this test vector, regardless of the fact that it uses different HTLCs from the other vectors? So the issue probably lies somewhere else, unless I'm missing something...

Reply:
Unfortunately, not. The HTLCs are used to calculate `to_local` and `to_remote` for the commitment transaction.

https://github.com/rust-bitcoin/rust-lightning/blob/9c9c88147e6a8fa637dc048f764ab8ecfea200d4/lightning/src/ln/channel.rs#L919-L925

So as the test vectors are written, `to_local`, `to_remote`, HTLCs, and fees must total the channel value (i.e., 10_000_000 sats). But for the new test vector they are 1_000 sats short.

Reply:
Got it, I'll change the HTLC amounts then to make it easier.

Reply:
Done in https://github.com/lightningnetwork/lightning-rfc/pull/539/commits/d178fb95907efabcec4e9c7a520e269a80eb15f8, let me know if that's easier for you.